File: pinyin-game-api/app/schemas/practice.py
"""练习与游戏 Schema。"""

from __future__ import annotations
from pydantic import BaseModel
from typing import List


class GameCardOut(BaseModel):
    """游戏宫格卡片（不暴露 pair_key 给前端判题，仅用于展示与本地配对 UX）。"""
    card_id: str
    question_id: int
    card_type: str  # hanzi | pinyin
    text: str


class GameDataOut(BaseModel):
    book_id: int
    book_title: str
    total: int
    cards: List[GameCardOut]


class AnswerItem(BaseModel):
    question_id: int
    user_pinyin: str


class PracticeSubmitRequest(BaseModel):
    book_id: int
    answers: List[AnswerItem]
    duration_seconds: int = 0


class PracticeRecordOut(BaseModel):
    id: int
    book_id: int
    book_title: str
    total_count: int
    correct_count: int
    accuracy: float
    duration_seconds: int

    class Config:
        from_attributes = True


class AnswerDetailOut(BaseModel):
    hanzi: str
    user_pinyin: str
    correct_pinyin: str
    is_correct: bool

File: pinyin-game-web/src/api/practice.ts
import request from '@/utils/request'

export interface AnswerItem {
  question_id: number
  user_pinyin: string
}

export interface SubmitResult {
  record_id: number
  book_title: string
  total_count: number
  correct_count: number
  accuracy: number
  duration_seconds: number
}

export interface RecordDetail {
  id: number
  book_id: number
  book_title: string
  total_count: number
  correct_count: number
  accuracy: number
  duration_seconds: number
  details: {
    hanzi: string
    user_pinyin: string
    correct_pinyin: string
    is_correct: boolean
  }[]
}

export function submitPractice(data: {
  book_id: number
  answers: AnswerItem[]
  duration_seconds: number
}) {
  return request.post<any, SubmitResult>('/web/practice/submit', data)
}

export function fetchRecord(recordId: number) {
  return request.get<any, RecordDetail>(`/web/practice/records/${recordId}`)
}
